### PR TITLE
feat: support keyword arguments in marker expressions

### LINKED ISSUE #12281: Support marker keyword arguments in marker expression (test selection via `-m`)
#### What's the problem this feature will solve?

It's possible to create custom markers like so:
```py
@pytest.mark.my_device_under_test
```

This marker can the be used to de/select tests via marker expressions, e.g.:
`pytest -m "not my_device_under_test"`

It is also possible to define keyword arguments for custom markers, e.g.:

```py
@pytest.mark.my_device_under_test(with_serial="1234")
```

Now, it would be amazing to select tests further depending on their keyword arguments + values, e.g.:

`pytest -m "my_device_under_test(with_serial="1234)"`

So only tests with that specific marker + that specific keyword argument+value would be run.

#### Describe the solution you'd like

Marker keyword argument selection in pytest's marker expression.

#### Alternative Solutions

- There was a `nose` test plugin, which supported something similar [here](https://github.com/nose-devs/nose/blob/master/nose/plugins/attrib.py). This could be ported to pytest but comes with it's own `attr` decorator.

- A pytest plugin could probably do this in an extra step as well, but would somehow need to disable pytest's internal marker selection plugin [here](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/mark/__init__.py#L261) as it would collide/fail otherwise.

#### Additional context
When skimming through the code I saw that @bluetech implemented a custom parsers some years ago. I'm not very familiar with the topic, as in how much effort it would be support such feature. I'd be interested to contribute though.

### PR DESCRIPTION
**Pull request by lovetheguitar**
Fixes #12281

With this change you can restrict a test run to only run tests matching one or multiple marker
keyword arguments
Only keyword argument matching is supported in marker expressions.
Only `int`, `str`, `bool` & `None` values are supported in marker expressions.

### CODE REVIEW

**Comment by webknjaz** [comment]
@lovetheguitar could you rebase the PR, fill out the PR description/commits, and compose a change note?

File: src/_pytest/mark/__init__.py [hunk 1]
```diff
@@ -181,7 +183,9 @@ def from_item(cls, item: "Item") -> "KeywordMatcher":
 
         return cls(mapped_names)
 
-    def __call__(self, subname: str) -> bool:
+    def __call__(self, subname: str, /, **kwargs: object) -> bool:
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:186]: `kwargs` is always a mapping, I don't think that using `object` for typing such things is customary, is it?

File: src/_pytest/mark/__init__.py [hunk 2]
```diff
@@ -211,24 +215,41 @@ def deselect_by_keyword(items: "List[Item]", config: Config) -> None:
...
+            return False
+
+        if not kwargs:
+            return True
 
-    def __call__(self, name: str) -> bool:
-        return name in self.own_mark_names
+        for mark in matches:
+            if all(
+                mark.kwargs.get(k, NOT_NONE_SENTINEL) == v for k, v in kwargs.items()
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:248]: Sentinel objects are typically compared though the identity operator, not equality (`is None`, `is not False` etc.)

So the following might be more correct:
```suggestion
                mark.kwargs.get(k, NOT_NONE_SENTINEL) is v for k, v in kwargs.items()
```

Comment (webknjaz) [reply src/_pytest/mark/__init__.py:248]: Or perhaps even
```suggestion
                v is NOT_NONE_SENTINEL or mark.kwargs.get(k, NOT_NONE_SENTINEL) == v for k, v in kwargs.items()
```

File: src/_pytest/mark/__init__.py [hunk 3]
```diff
@@ -211,24 +215,41 @@ def deselect_by_keyword(items: "List[Item]", config: Config) -> None:
...
     @classmethod
     def from_item(cls, item: "Item") -> "MarkMatcher":
-        mark_names = {mark.name for mark in item.iter_markers()}
-        return cls(mark_names)
+        mark_name_mapping = collections.defaultdict(list)
+        for mark in item.iter_markers():
+            mark_name_mapping[mark.name].append(mark)
+        return cls(mark_name_mapping)
+
+    def __call__(self, name: str, /, **kwargs: object) -> bool:
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:239]: Same here: the typing should be clarified, it seems.

File: src/_pytest/mark/__init__.py [hunk 4]
```diff
@@ -181,7 +183,9 @@ def from_item(cls, item: "Item") -> "KeywordMatcher":
 
         return cls(mapped_names)
 
-    def __call__(self, subname: str) -> bool:
+    def __call__(self, subname: str, /, **kwargs: object) -> bool:
+        if kwargs:
+            raise UsageError("Keyword expressions do not support call parameters.")
```
[the remark this answers is not in the document]
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py:189]: Nope, we need a `UsageError` so pytest handles the CLI error properly. Otherwise we'd get lovely `INTERNALERROR>`s.

File: src/_pytest/mark/__init__.py (hunk 2 above)
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py]: I do not really care about the sentinel object, I just need it to not screw up comparing e.g. `None` (when I did not use any). The other values I need to compare have to be compared by value.

**Comment by lovetheguitar** [comment]
@RonnyPfannschmidt Could you please have a look? :)

File: src/_pytest/mark/__init__.py (hunk 1 above)
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py]: Switched to `str | int | bool | None` for now.

**Comment by lovetheguitar** [comment]
@webknjaz https://readthedocs.org/projects/pytest/builds/24773128/ does the preview fail to link that or did I reference it wrongly?

File: changelog/12281.feature.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Added support for keyword matching in marker expressions.
+
+Now tests can be selected by marker keyword arguments.
+Supported values are `int`, (unescaped) `str`, `bool` & `None`.
+
+See :ref:`marker examples <marker_keyword_expression_example>` for more information.
+
+-- by :user:`lovetheguitar`.
```
Comment (webknjaz) [inline changelog/12281.feature.rst:8]:
```suggestion
-- by :user:`lovetheguitar`
```

File: AUTHORS [hunk 6]
```diff
@@ -244,6 +244,7 @@ Levon Saldamli
 Lewis Cowles
 Llandy Riveron Del Risco
 Loic Esteve
+lovetheguitar
```
Comment (webknjaz) [inline AUTHORS:247]: Are you sure you don't want the name here?

File: changelog/12281.feature.rst [hunk 7]
```diff
@@ -0,0 +1,8 @@
+Added support for keyword matching in marker expressions.
+
+Now tests can be selected by marker keyword arguments.
+Supported values are `int`, (unescaped) `str`, `bool` & `None`.
```
Comment (webknjaz) [inline changelog/12281.feature.rst:4]:
```suggestion
Supported values are :class:`int`, (unescaped) :class:`str`, :class:`bool` & :data:`None`.
```

File: doc/en/example/markers.rst [hunk 8]
```diff
@@ -71,6 +73,28 @@ Or the inverse, running all tests except the webtest ones:
...
+    rootdir: /home/sweet/project
+    collecting ... collected 4 items / 3 deselected / 1 selected
+
+    test_server.py::test_something_quick PASSED                          [100%]
+
+    ===================== 1 passed, 3 deselected in 0.12s ======================
+
+.. note:: Only keyword argument matching is supported in marker expressions.
+
+.. note:: Only `int`, (unescaped) `str`, `bool` & `None` values are supported in marker expressions.
```
Comment (webknjaz) [inline doc/en/example/markers.rst:96]: Single backticks are for Markdown. In RST they often won't behave as you think. Either use double backticks or declare explicit roles:
```suggestion
.. note:: Only :class:`int`, (unescaped) :class:`str`, :class:`bool` & :data:`None` values are supported in marker expressions.
```

File: doc/en/how-to/usage.rst [hunk 9]
```diff
@@ -76,11 +76,19 @@ Specifying a specific parametrization of a test:
 
 **Run tests by marker expressions**
 
+To run all tests which are decorated with the ``@pytest.mark.slow`` decorator:
```
Comment (webknjaz) [inline doc/en/how-to/usage.rst:79]:
```suggestion
To run all tests which are decorated with the ``@pytest.mark.slow`` decorator, execute the following command:
```

File: doc/en/how-to/usage.rst [hunk 10]
```diff
@@ -76,11 +76,19 @@ Specifying a specific parametrization of a test:
...
+
 .. code-block:: bash
 
     pytest -m slow
 
-Will run all tests which are decorated with the ``@pytest.mark.slow`` decorator.
+
+To run all tests which are decorated with the annotated ``@pytest.mark.slow(phase=1)`` decorator,
+with the `phase` keyword argument set to `1`:
```
Comment (webknjaz) [inline doc/en/how-to/usage.rst:87]:
```suggestion
with the ``phase`` keyword argument set to ``1``, use:
```

File: src/_pytest/mark/__init__.py [hunk 11]
```diff
@@ -218,17 +223,29 @@ class MarkMatcher:
...
-    def __call__(self, name: str) -> bool:
-        return name in self.own_mark_names
+    def __call__(self, name: str, /, **kwargs: str | int | bool | None) -> bool:
+        if not (matches := self.own_mark_name_mapping.get(name, [])):
+            return False
+
+        if not kwargs:
+            return True
+
+        for mark in matches:
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:244]: Is this line covered? Where do matches come from? Wouldn't walrus get destroyed outside its if-block?

File: src/_pytest/mark/__init__.py [hunk 12]
```diff
@@ -218,17 +223,29 @@ class MarkMatcher:
...
+    def from_markers(cls, markers: Iterable[Mark]) -> MarkMatcher:
+        mark_name_mapping = collections.defaultdict(list)
+        for mark in markers:
+            mark_name_mapping[mark.name].append(mark)
+        return cls(mark_name_mapping)
 
-    def __call__(self, name: str) -> bool:
-        return name in self.own_mark_names
+    def __call__(self, name: str, /, **kwargs: str | int | bool | None) -> bool:
+        if not (matches := self.own_mark_name_mapping.get(name, [])):
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:238]: This var isn't actually used within the block so there's no need to have walrus here:
```suggestion
        if not self.own_mark_name_mapping.get(name, []):
```

File: doc/en/how-to/usage.rst (hunk 9 above)
Comment (lovetheguitar) [reply doc/en/how-to/usage.rst:79]: The suggestion is incoherent with previous examples, leaving as is.

File: doc/en/how-to/usage.rst (hunk 10 above)
Comment (lovetheguitar) [reply doc/en/how-to/usage.rst]: ✅

File: src/_pytest/mark/__init__.py (hunk 12 above)
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py:238]: Yes it is on L244.

File: src/_pytest/mark/__init__.py (hunk 11 above)
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py:244]: Yes, from above, nope.

File: doc/en/example/markers.rst (hunk 8 above)
Comment (lovetheguitar) [reply doc/en/example/markers.rst]: Yep, using markdown always, ...changed to double back ticks everywhere.

File: changelog/12281.feature.rst (hunk 5 above)
Comment (lovetheguitar) [reply changelog/12281.feature.rst:8]: 😆 copied that one over from one of the latest fragments added during the sprint.

File: changelog/12281.feature.rst (hunk 7 above)
Comment (lovetheguitar) [reply changelog/12281.feature.rst]: Double backticks it is.

File: AUTHORS (hunk 6 above)
Comment (lovetheguitar) [reply AUTHORS:247]: ✅ Fine like it is.

**Review by lovetheguitar — COMMENTED** [review]
Should be good to go @RonnyPfannschmidt. 🚀 🤞

**Review by RonnyPfannschmidt — APPROVED** [review]

File: changelog/12281.feature.rst (hunk 5 above)
Comment (webknjaz) [reply changelog/12281.feature.rst:8]: That's different with a single sentence vs. a paragraph and a and a separate byline.

File: doc/en/example/markers.rst (hunk 8 above)
Comment (webknjaz) [reply doc/en/example/markers.rst]: It's usually better to use explicit roles, though.

File: src/_pytest/mark/__init__.py (hunk 12 above)
Comment (webknjaz) [reply src/_pytest/mark/__init__.py:238]: Ah, okay, I got confused with how it works.

File: src/_pytest/mark/__init__.py [hunk 13]
```diff
@@ -218,17 +223,29 @@ class MarkMatcher:
...
+            return False
+
+        if not kwargs:
```
Comment (RonnyPfannschmidt) [inline src/_pytest/mark/__init__.py:241]: This case cannot happen, I'll fix it when I get back to the computer, but feel free to beat me to it

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]

File: changelog/12281.feature.rst [hunk 14]
```diff
@@ -0,0 +1,8 @@
+Added support for keyword matching in marker expressions.
+
+Now tests can be selected by marker keyword arguments.
+Supported values are ``int``, (unescaped) ``str``, ``bool`` & ``None``.
```
Comment (webknjaz) [inline changelog/12281.feature.rst:4]:
```suggestion
Supported values are :class:`int`, (unescaped) :class:`str`, :class:`bool` & :data:`None`.
```

File: doc/en/example/markers.rst [hunk 15]
```diff
@@ -71,6 +73,28 @@ Or the inverse, running all tests except the webtest ones:
...
+    rootdir: /home/sweet/project
+    collecting ... collected 4 items / 3 deselected / 1 selected
+
+    test_server.py::test_something_quick PASSED                          [100%]
+
+    ===================== 1 passed, 3 deselected in 0.12s ======================
+
+.. note:: Only keyword argument matching is supported in marker expressions.
+
+.. note:: Only ``int``, (unescaped) ``str``, ``bool`` & ``None`` values are supported in marker expressions.
```
Comment (webknjaz) [inline doc/en/example/markers.rst:96]:
```suggestion
.. note:: Only :class:`int`, (unescaped) :class:`str`, :class:`bool` & :data:`None` values are supported in marker expressions.
```

File: src/_pytest/mark/__init__.py [hunk 16]
```diff
@@ -218,17 +223,29 @@ class MarkMatcher:
...
+            return True
```
Comment (webknjaz) [inline src/_pytest/mark/__init__.py:242]:
```suggestion
```

**Review by webknjaz — APPROVED** [review]

**Comment by RonnyPfannschmidt** [comment]
awesome - @lovetheguitar thanks for seeing this tough and @webknjaz thanks for sorting out the last details to land this

File: doc/en/example/markers.rst [hunk 17]
```diff
@@ -71,6 +73,28 @@ Or the inverse, running all tests except the webtest ones:
...
     ===================== 3 passed, 1 deselected in 0.12s ======================
 
+.. _`marker_keyword_expression_example`:
+
+Additionally, you can restrict a test run to only run tests matching one or multiple marker
+keyword arguments, e.g. to run only tests marked with ``device`` and the specific ``serial="123"``:
+
+.. code-block:: pytest
+
+    $ pytest -v -m 'device(serial="123")'
```
Comment (The-Compiler) [inline doc/en/example/markers.rst:83]: Single quotes won't work on Windows, which is why we use double quotes above.

File: doc/en/how-to/usage.rst [hunk 18]
```diff
@@ -76,11 +76,19 @@ Specifying a specific parametrization of a test:
...
     pytest -m slow
 
-Will run all tests which are decorated with the ``@pytest.mark.slow`` decorator.
+
+To run all tests which are decorated with the annotated ``@pytest.mark.slow(phase=1)`` decorator,
+with the ``phase`` keyword argument set to ``1``:
+
+.. code-block:: bash
+
+    pytest -m slow(phase=1)
```
Comment (The-Compiler) [inline doc/en/how-to/usage.rst:91]: Needs quotes to work.

File: testing/test_mark.py [hunk 19]
```diff
@@ -233,6 +233,54 @@ def test_two():
     assert passed_str == expected_passed
 
 
+@pytest.mark.parametrize(
+    ("expr", "expected_passed"),
+    [  # TODO: improve/sort out
```
Comment (The-Compiler) [inline testing/test_mark.py:238]: Leftover TODO?

File: testing/test_mark_expression.py [hunk 20]
```diff
@@ -195,3 +200,119 @@ def test_valid_idents(ident: str) -> None:
...
+        ("""mark(var="')""", 'closing quote """ is missing'),
+        ("""mark(var='")""", """closing quote "'" is missing"""),
+        (
+            r"mark(var='\hugo')",
+            r'escaping with "\\" not supported in marker expression',
+        ),
+        ("mark(empty_list=[])", r'unexpected character/s "\[\]"'),
+    ),
+)
+def test_invalid_kwarg_name_or_value(  # TODO: move to `test_syntax_errors` ?
```
Comment (The-Compiler) [inline testing/test_mark_expression.py:233]: Leftover TODO?

File: testing/test_mark_expression.py [hunk 21]
```diff
@@ -195,3 +200,119 @@ def test_valid_idents(ident: str) -> None:
...
+        ("builtin_matchers_mark(y=True)", False),
+        ("builtin_matchers_mark(y=False)", True),
+        ("builtin_matchers_mark(z=None)", True),
+        ("builtin_matchers_mark(z=False)", False),
+        ("builtin_matchers_mark(z=True)", False),
+        ("builtin_matchers_mark(z=0)", False),
+        ("builtin_matchers_mark(z=1)", False),
+    ),
+)
+def test_builtin_matchers_keyword_expressions(  # TODO: naming when decided
```
Comment (The-Compiler) [inline testing/test_mark_expression.py:292]: Leftover TODO?

**Review by The-Compiler — COMMENTED** [review]
I think this is amazing work, and it's great how many tests there are and corner cases there are handled!

I'm really unhappy how this was merged in a seemingly unfinished state (TODO comments in various places), while we were eating dinner together, 2 hours after the final version was pushed. I found a couple of pretty obvious things from a cursory review, and now I'll either need to do a follow-up PR to pick up the lose ends, or someone else needs to go back to opening a new PR and all that additional work.

Also, @bluetech wrote the parser in the first place if memory serves me right, so I feel like it'd be appropriate to give him a chance to look at this, no?

File: testing/test_mark_expression.py (hunk 20 above)
Comment (lovetheguitar) [reply testing/test_mark_expression.py:233]: Yep, sorry, forgot to remove that one, talked about it with Ronny to keep it separate as is.

File: testing/test_mark_expression.py (hunk 21 above)
Comment (lovetheguitar) [reply testing/test_mark_expression.py:292]: Same for that one. 🙃

**Comment by lovetheguitar** [comment]
> I think this is amazing work, and it's great how many tests there are and corner cases there are handled!
> 
> I'm really unhappy how this was merged in a seemingly unfinished state (TODO comments in various places), while we were eating dinner together, 2 hours after the final version was pushed. I found a couple of pretty obvious things from a cursory review, and now I'll either need to do a follow-up PR to pick up the lose ends, or someone else needs to go back to opening a new PR and all that additional work.
> 
> Also, @bluetech wrote the parser in the first place if memory serves me right, so I feel like it'd be appropriate to give him a chance to look at this, no?

@The-Compiler Sure thing, I was also thinking about mentioning @bluetech, since, yes he wrote that part and he commented also in the issue.

Feel free to be nitpicky, I'm motivated to learn the customs properly and do the right thing (TM). 😄

File: changelog/12281.feature.rst [hunk 22]
```diff
@@ -0,0 +1,8 @@
+Added support for keyword matching in marker expressions.
+
+Now tests can be selected by marker keyword arguments.
+Supported values are :class:`int`, (unescaped) :class:`str`, :class:`bool` & :data:`None`.
+
```
Comment (bluetech) [inline changelog/12281.feature.rst:5]: It would be nice to add an example here, for people skimming the changelog without clicking through to the link.

File: src/_pytest/mark/expression.py [hunk 23]
```diff
@@ -5,7 +5,8 @@
 expression: expr? EOF
 expr:       and_expr ('or' and_expr)*
 and_expr:   not_expr ('and' not_expr)*
-not_expr:   'not' not_expr | '(' expr ')' | ident
+not_expr:   'not' not_expr | '(' expr ')' | ident ( '(' name '=' value ( ', ' name '=' value )*  ')')*
```
Comment (bluetech) [inline src/_pytest/mark/expression.py:8]: - Would remove the space after `ident (` for consistency.
- The `*` (zero-or-more) at the end of this line should be `?` (zero-or-one), I think.
- The definitions of the `name` and `value` productions are missing.

File: src/_pytest/mark/expression.py [hunk 24]
```diff
@@ -5,7 +5,8 @@
...
+
 ident:      (\w|:|\+|-|\.|\[|\]|\\|/)+
 
 The semantics are:
```
Comment (bluetech) [inline src/_pytest/mark/expression.py:12]: Let's describe here the semantics of `ident(...)`.

File: src/_pytest/mark/expression.py [hunk 25]
```diff
@@ -43,6 +47,9 @@ class TokenType(enum.Enum):
     NOT = "not"
     IDENT = "identifier"
     EOF = "end of input"
+    EQUAL = "="
+    STRING = "str"
```
Comment (bluetech) [inline src/_pytest/mark/expression.py:51]: "str" might be confusing, I suggest "string literal"

**Review by bluetech — COMMENTED** [review]
Very nice work, I wouldn't have done it better myself :)

I left some comments.

**Comment by webknjaz** [comment]
Too bad that it got auto-merged earlier, than intended. I knew that Ronny didn't want to make the process too difficult for a first-time contributor so I fixed a few concerns myself and hit `Approve` to replace the blocking review, assuming that Ronny did check the other things. And I addressed Ronny's comment and dismissed his review, communicating with him offline. This was in line with Ronny's plan and the auto-merge that he'd set earlier (it wasn't auto-reset after my commits). I should've instead dismissed my review so it'd not be a blocker and not hit approve, as there were parts I didn't really inspect deeply.

@lovetheguitar since you stated you wanted to do better, here are some action items that you could perform (with an attempt to explain why they are important):
1) Make a follow-up PR or several of them, addressing the comments that Ran and Florian made (I suppose, this one is rather obvious);
2) When you see examples of things similar to what you're adding, it's usually best to follow the surrounding style — this avoids additional mental overhead for the reviewers since the pre-existing pieces have been battle-tested and any customizations would need to be inspected from various angles and might be missed by some reviewers (I missed it, for example, and didn't even know it was significant);
3) I usually recommend not closing/dismissing the separate inline review threads started by someone else for the following reasons:
   * only the thread creator knows if you actually addressed their concerns — you can guess but in general, the only way of really knowing is to let them declare the thing solved. A review is a multi-sided conversation and not someone throwing in instructions to be either followed or dismissed silently — be verbose about what's going on and let others get onto the same page with you;
   * clicking the `Resolve conversation` button silently adds mental overhead for the creator since it doesn't generate any notifications and just becomes collapsed; It then requires someone to click on all the collapsed threads, trying to figure out where that old context exists and attempt to judge whether it was solved and guess why it disappeared as the reason is not always obvious and see the previous point for why it might not actually be resolved. It creates more round-trips, confusion, and frustration. I recommend just re-asking if you feel like it might be resolved or waiting;
   * in the previous point, I mentioned that no-notification disappearance of threads. To solve this point, if you're like really sure you can decide that it's resolved by yourself (for example, when you followed what was requested to the letter), I suggest writing a summary/notification comment about this action — I typically explain why/how the thing was resolved and add explicit text like "Marking this thread as resolved." This lets other people actually find it by following their email notifications and registering that they don't need to follow that conversation anymore. I've also seen great pull request authors including commit links there — in case they didn't click on the `Commit suggestion` button and re-implemented it.

That point (3) is something that happened here — I suggested a few correct fixes and extra explanations in some places. It was possible to click one button on the GH UI and get them in. However, they ended up being dismissed for a mysterious reason that I still have no idea of. There was a comment about choosing to use an alternative (worse) syntax option but the motivation was never stated. So I still had to just add those things manually which prolonged the PR cycle — because of lack of communication.
And with the other suggestion, it was dismissed with an incorrect perception, forcing me to dig the thread from the pile of others and unresolve it.

To summarize, the PR was good in general, but a few things were missed during the review and the transparency could be better (not just on the PR author's side, though).

File: testing/test_mark.py (hunk 19 above)
Comment (lovetheguitar) [reply testing/test_mark.py:238]: Preliminarily dropped.

File: src/_pytest/mark/__init__.py (hunk 13 above)
Comment (lovetheguitar) [reply src/_pytest/mark/__init__.py]: That was an early return to keep the logic for "non-keyword" markers fast/similar to before.

The case happens every time a "non-keyword" marker is looked up https://github.com/pytest-dev/pytest/blob/f75e3fe63f8d22422393888bd5b9094cab35f522/src/_pytest/mark/expression.py#L271-L272

**Comment by lovetheguitar** [comment]
> Very nice work, I wouldn't have done it better myself :)
> 
> I left some comments.

Thanks for the fast review @bluetech. I tried to address the open points in #12523.
Looking forward to some feedback.

File: changelog/12281.feature.rst (hunk 22 above)
Comment (webknjaz) [reply changelog/12281.feature.rst:5]: @lovetheguitar I think, it's nice to have examples in such places, but perhaps only small ones. I like Ran's idea. If you make a PR for that, we could discuss something more concrete there, with dedicated review and polishing.

Comment (webknjaz) [reply changelog/12281.feature.rst:5]: @bluetech I believe the previous comment of @lovetheguitar was addressed to you, eh?
